Development of documentation (#66)
ARM-software/connectedhomeip: docs/examples/openiotsdk_examples.md

emulated target through the
[Arm FVP model for the Corstone-300 MPS3](https://developer.arm.com/downloads/-/arm-ecosystem-fvps).

> The [list](../../examples/platform/openiotsdk/supported_examples.md) of

Author: The idea is to share this file between `openiotsdk_example.sh` script, VSCode tasks and documentation. Only one modification when you want to add a new example.

currently supported Matter examples.

You can use these examples as a reference for creating your own applications.

## Environment setup

The VSCode devcontainer has all the dependencies pre-installed. Using the VSCode
devcontainer is the recommended way to interact with the Open IoT SDK port of
the Matter Project. Please read this
[VSCode development guide](../VSCODE_DEVELOPMENT.md) for more information.

Before building the examples, check out the Matter repository and sync Open IoT
SDK submodules using the following command:

```
scripts/checkout_submodules.py --shallow --recursive --platform openiotsdk
```

The next step is bootstrapping the source tree (only once). This installs
Pigweed (CIPD and Python packages) components inside your environment.

To bootstrap:

**using CLI**

```
$ bash scripts/bootstrap.sh
```

**using VSCode tasks**

```
Command Palette (F1)
=> Tasks: Run Task

> => Bootstrap

Author: We can't use the <kbd> in code block

Connection Plugin for Fast Model. More details
[GDBRemoteConnection](https://developer.arm.com/documentation/100964/1116/Plug-ins-for-Fast-Models/GDBRemoteConnection).

The `Fast Models FVP` add-on package can be downloaded from the ARM developer
website [Fast models](https://developer.arm.com/downloads/-/fast-models). After
login in to the `ARM developer` platform search for `Fast Models`, choose
`Fast Models (FM000A)` on the list of results, then choose the revision
`r11p16-16rel0` and download the
`Third Party Add-ons for Fast Models 11.16 (Linux)` package. Then unpack the
package in the selected location on the host machine.

> The next step is to add the GDB Remote Connection Plugin to your development
> environment:

Reviewer: ```suggestion
Now you should add the GDB Remote Connection Plugin to your development
environment:
```

        environment by mounting it into the
        `/opt/FastModels_ThirdParty_IP_11-16_b16_Linux64` directory in the
        container. Add a volume bound to this directory
        [Add local file mount](https://code.visualstudio.com/remote/advancedcontainers/add-local-file-mount).

        You can edit the `.devcontainer/devcontainer.json` file, for example:

        ```
        ...
        "mounts": [ ...
        "source=/opt/FastModels_ThirdParty_IP_11-16_b16_Linux64,target=/opt/FastModels_ThirdParty_IP_11-16_b16_Linux64,type=bind,consistency=cached"
        ... ],
        ...
        ```

        Or if you launch the Docker container directly from CLI, use the above
        arguments with `docker run` command:

        ```
        docker run ... --mount type=bind,source=/opt/FastModels_ThirdParty_IP_11-16_b16_Linux64,target=/opt/FastModels_ThirdParty_IP_11-16_b16_Linux64 ...
        ```

    -   install the Fast Model Extension package via setup script inside Docker
        container:

        ```
        ${MATTER_ROOT}/scripts/setup/openiotsdk/debugging_setup.sh
        ```

    -   the GDB Remote Connection Plugin should be visible in

>         `opt/FastModelsPortfolio_11.16/plugins/Linux64_GCC-9.3` directory.

Reviewer: ```suggestion
        `/opt/FastModelsPortfolio_11.16/plugins/Linux64_GCC-9.3` directory.
```

>         -   For `VScode devcontainer` users the `FAST_MODEL_PLUGINS_PATH`
>             environment variable is already created and points to the correct

Reviewer: ```suggestion
        -   For `VScode devcontainer` use the environment variable `FAST_MODEL_PLUGINS_PATH`
             to point to the correct
```

[Open IoT SDK build script](../../scripts/examples/openiotsdk_example.sh)
provides the `-K,--kvsfile` option to use the persistence options listed above.

### Crypto backend

Open IoT SDK port supports two crypto backend implementations:

-   [Mbed TLS](../guides/openiotsdk_platform_overview.md#mbed-tls) - it's the
    default option
-   [PSA crypto service](https://tf-m-user-guide.trustedfirmware.org/integration_guide/services/tfm_crypto_integration_guide.html)
    from the

>     [TF-M](../guides/openiotsdk_platform_overview.md#trusted-firmware-m)

Reviewer: ```suggestion
    [TrustedFirmware-M (TF-M)](../guides/openiotsdk_platform_overview.md#trusted-firmware-m)
```

    component

> The `CONFIG_CHIP_CRYPTO` CMake variable is used to set the crypto backend type.
> 
> Set the variable `CONFIG_CHIP_CRYPTO` to `mbedtls` to add `Mbed TLS` support to
> your application. You can put it inside the main application `CMakeLists.txt`
> file:
> 
> ```
> set(CONFIG_CHIP_CRYPTO mbedtls)
> ```
> 
> or add as a Cmake command-line parameter:
> 
> ```
> cmake -G <...> -DCONFIG_CHIP_CRYPTO=mbedtls <...>
> ```
> 
> To switch to the `TF-M PSA crypto` set the variable `CONFIG_CHIP_CRYPTO` to
> `psa`. You can put it inside the main application `CMakeLists.txt` file:
> 
> ```
> set(CONFIG_CHIP_CRYPTO psa)
> ```
> 
> or add as a Cmake command-line parameter:
> 
> ```
> cmake -G <...> -DCONFIG_CHIP_CRYPTO=psa <...>
> ```
> 
> This option is set to use the
> [PSA Cryptography API](https://armmbed.github.io/mbed-crypto/html/) for crypto
> operations in the Matter stack.
> 

Reviewer: ```suggestion
The CMake variable `CONFIG_CHIP_CRYPTO` controls how cryptographic operations are implemented in Matter. It accepts two values:
- `mbedtls`: Use Mbed TLS for Cryptographic operations.
- `psa`: Use [PSA Cryptography API](https://armmbed.github.io/mbed-crypto/html/) for crypto
operations.

This variable can be set in the main application `CMakeLists.txt`:

```
set(CONFIG_CHIP_CRYPTO <mbedtls | psa>)
```

The variable can also be defined with CMake CLI:

```
cmake -G <...> -DCONFIG_CHIP_CRYPTO=<mbedtls | psa> <...>
```
```

**NOTES**

The `TF-M PSA crypto` option requires enabling [TF-M](#trusted-firmware-m)
support.

### IoT Socket API

By default, the [LwIP](../guides/openiotsdk_platform_overview.md#lwip) is used
as a system layer and communication endpoints implementation.

There is an option to switch to
[IoT Socket](../guides/openiotsdk_platform_overview.md#iot-socket) usage for
these Matter components. Set the variable
`CONFIG_CHIP_OPEN_IOT_SDK_USE_IOT_SOCKET` to `YES` to add the `IoT Socket API`
support in your application. You can put it inside the main application
`CMakeLists.txt` file:

```
set(CONFIG_CHIP_OPEN_IOT_SDK_USE_IOT_SOCKET YES)
```

or add as a Cmake command-line parameter:

```
cmake -G <...> -DCONFIG_CHIP_OPEN_IOT_SDK_USE_IOT_SOCKET=YES <...>
```

This option is set to use the `IoT Socket` version of a system layer and
communication endpoints in the application. The
[CMSIS-RTOS2 API](../guides/openiotsdk_platform_overview.md#rtos) is required in
this type of system layer. It's supported by default in `Open IoT SDK` port.

## Building

> You can build an example by using a VSCode task or by calling the build script
> directly from the command line.

Reviewer: ```suggestion
You can build examples using the dedicated VSCode task or by calling directly the build script
from the command line.
```

## Running

The application runs in the background and opens a telnet session. The telnet
client connects to the port used by the `FVP`. When the telnet process is

> terminated it will also terminate the `FVP` instance.

Reviewer: ```suggestion
terminated it also terminates the `FVP` instance.
```

Before debugging ensure the following:

1. The debug environment is correctly setup:
   [debugging setup](#debugging-setup).

2. The example is compiled with debug symbols enabled:

    For CLI:

    ```
    ${MATTER_ROOT}/scripts/examples/openiotsdk_example.sh -d true <example name>
    ```

    For the VSCode task:

    ```
    => Use debug mode (true)
    ```

3. The test network is correctly setup (if required):

>    [networking setup](#networking-setup).

Reviewer: ```suggestion
   see [networking setup](#networking-setup).
```

=> Command (restart)
```

```
Command Palette (F1)
=> Tasks: Run Task
=> Test Open IoT SDK example
=> Network namespace:  OIStestns
=> Network interface name: OIStesttap
=> Example application to use (lock-app)
```

### Build lock-app example with mbedtls crypto backend and lwip support in debug mode and debug it in the network namespace using the VSCode task

```
Command Palette (F1)
=> Tasks: Run Task
=> Build Open IoT SDK example
=> Use debug mode (true)
=> Use LwIP debug logs (false)
=> Crypto algorithm to use (mbedtls)
=> Socket API to use (lwip)
=> Example application to use (lock-app)
```

```
Command Palette (F1)
=> Tasks: Run Task
=> Setup Open IoT SDK network
=> Network namespace name: OIStest
=> Command (restart)
```

```
1. Click 'Run and Debug' from the primary side menu or press (Ctrl+Shift+D)
2. Select 'Debug Open IoT SDK example application' from the drop down list
3. Click 'Start Debugging'(green triangle) or press (F5)
4. Enter:
=> lock-app
=> GDB target address: 10.200.1.2
=> Network namespace: OIStestns
=> Network interface name: OIStesttap
=> Example application to use (lock-app)
```

Use debug controls to debug the application.

## Add new example

> This chapter describes the details to add a new Matter example for the Open IoT
> SDK platform. Use the same `example_name` in each step below.

Reviewer: ```suggestion
This chapter describes how to add a new Matter example based on Open IoT
SDK platform. 

In the description below we use the placeholder `example_name` as the name of the example to create. Replace it with the name of your example. 
```

### Files structure

A new example should be put into `examples/<example_name>/openiotsdk` directory.
It should contain:

-   application source files and headers in the `main` sub-directory
-   application `CMakeLists.txt` file
-   `.gitignore` file with with all sources to skip
-   `README.md` file with example description
-   additional directories with required configuration for used components. Use
    `component_name-config` pattern, e.g `freertos-config`

### Target name

A new application target name should be created with
`chip-openiotsdk-<example_name>-example(_ns)` pattern. The `_ns` suffix is
required for [TF-M applications](#trusted-firmware-m).

Example:

```
set(APP_TARGET chip-openiotsdk-new-example-example_ns)
```

### Example tools

Add a new example name to the

> [supported examples list](../../examples/platform/openiotsdk/supported_examples.md).

Reviewer: I think we should mention that this file must be amended as well.

Author: What do you mean? Which file exactly?
